File: Data Structures & Algorithms/range-sum-of-bst/submission-1.py
"""
Understand:
Provided a tree and [low, high], return the sum of nodes that are within low/high
Example1, (3,8), 5+3+8+4+7=27
Example2, (2,4), 4+3+2=9
All nodes' values are unique

Match:
BFS, Intervals

Plan:
Do a BFS from root, add the value of node if it's within low/high values inclusively
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if root is None:
            return 0

        result = 0
        s = []
        s.append(root)

        while s:
            node = s.pop()
            
            if node is None:
                continue
            if low <= node.val <= high:
                result += node.val
            s.append(node.left)
            s.append(node.right)
        return result

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;

        auto dfs = [&](this auto&& self, TreeNode* node, int depth) -> void {
            // Return if current node is empty
            if (node == nullptr) {
                return;
            }

            // Push an empty container for this level if result's size equals depth
            if (static_cast<int>(result.size()) == depth) {
                result.push_back(vector<int>{});
            }

            // Push current node's value into result[depth]
            result[depth].push_back(node->val);

            // Recursive call with left/right nodes and (depth+1)
            self(node->left, depth + 1);
            self(node->right, depth + 1);
        };

        dfs(root, 0);
        return result;
    }
};
